products/bridge/smart-contracts/contracts/Relayer.sol

```diff
@@ -107,9 +110,46 @@ contract Relayer {
             call,
             readonly,
             callback,
-            nonces[msg.sender]
+            _nonces[msg.sender]
         );
-        return ++nonces[msg.sender];
+        return ++_nonces[msg.sender];
+    }
+
+    // TODO: remove later, used for testing
+    function warmup() external {
+        ++_refund[msg.sender];
+    }
+
+    function topUpGas(address target) external payable {
+        _gas[target] += msg.value;
+    }
+
+    function refundGas() external {
+        uint amount = _refund[msg.sender];
+        _refund[msg.sender] = 0;
+        payable(msg.sender).transfer(amount);
+    }
+
+    function refund(address x) external view returns (uint) {
+        return _refund[x];
+    }
+
+    function gas(address x) external view {
+        _gas[x];
+    }
+
+    modifier trackGas(address caller) {
+        uint gasStart = gasleft();
+        require(
+            _gas[caller] >= block.gaslimit * tx.gasprice,
+            "Insufficient gas to cover limit"
+        );
+        _;
+        // 27603 = 21000 + 3 + 6600
+        gasStart += 27603 + 16 * (msg.data.length - 4);
```

Comment:
What do you think of this estimation @DrZoltanFazekas ?
6600 is the number ive arrived to when i was testing dispatch with a call of 0 arguments, and 1 signature. This would be the theoretical minimum configuration for when such a call happens.
Also, Ive multiplied all bytes by 16 assuming they are non-zero. Hence the more calldata the larger the tip. Hence why the configuration with smallest calldata is the one to measure how much the baseline should cost

Reply:
Imo this approach is totally fine for now. I've started debugging some example code to see where exactly gas is being consumed depending on the size of the inputs. If I manage to find a formula that approximates the real gas usage more precisely and is not more expensive to compute than the tip, we can replace it in the code above.